### PR TITLE
[Internal] Engineering: Refactors code to reduce technical debt

### PR DESCRIPTION
**Pull request by marodev**
## Description
This pull request is _very simple_ and aims to reduce the technical debt. It specifically targets the "Duplicate casts should not be made" [rule](https://rules.sonarsource.com/csharp/RSPEC-3247) in SonarQube.

**Update:**

As pointed out by @ealsur, the SonarQube fixes touched OSS files that probably should't be modified. To make this contribution more valuable, I fixed the following ReSharper issues:

- "Replace if statement with null-propagating code" [rule](https://www.jetbrains.com/help/resharper/UseNullPropagation.html)
- "Merge sequential checks in && or || expressions"  [rule](https://www.jetbrains.com/help/resharper/MergeSequentialChecks.html)

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/ResourceThrottleRetryPolicy.cs [hunk 1]
```diff
@@ -48,9 +48,8 @@ internal sealed class ResourceThrottleRetryPolicy : IDocumentClientRetryPolicy
             Exception exception,
             CancellationToken cancellationToken)
         {
-            if (exception is DocumentClientException)
+            if (exception is DocumentClientException dce)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ResourceThrottleRetryPolicy.cs:51]: This is the only file that is SDK specific, the Antlr files are coming from a referenced Antlr open source project, I don't think we can modify them?

Comment (marodev) [reply Microsoft.Azure.Cosmos/src/ResourceThrottleRetryPolicy.cs:51]: Yes, you're probably right! I should've put my glasses on.

**Review by j82w — APPROVED** [review]

**Review by ealsur — APPROVED** [review]

**Pull request merged**